### PR TITLE
Feature/profiles : Profile Creation Screen On First Sign In

### PR DESCRIPTION
**Pull request by JanStaszewiczEPFL**
# Profile Creation Integration

## Changes Made

**Summary :** adding profile creation on first sign in and implementing it into app flow.

### Integrated Profile creation screen

- Added a `ProfileCreationScreen` composable that allows the user to customize profile on first sign in. See file `ProfileCreation.kt`.
- The screen was implemented into the app flow in the `MainActivity.kt` and in the navigation system.
- The SignIn.kt file was changed to integrate navigation to the `ProfileCreationScreen` if there was no `IceBreakrr` account for the user. 
- The Firebase FCM token generation was moved from `SignIn.kt` to the `ProfileCreationScreen` for this implementation. 

### Date Handling Standardization

- Created a `DateUtils` file in the `utils` folder to provide some basic date handling functions.

### Test Coverage

- Implemented tests for `ProfileCreation.kt`.
- Implemented tests for `DateUtils.kt`.

image

### CODE REVIEW

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        description.value.isNotBlank() &&
+        selectedGender != null
+  }
+
+  // Create DatePicker
+  val datePicker =
+      DatePickerDialog(
+          context,
+          { _, year, month, dayOfMonth ->
+            val newCalendar =
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:104]: You use twice the same instance of calendar
Copy it into one variable and reuse the variable

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+  val user = profilesViewModel.selectedProfile.collectAsState().value
+
+  val selectedTags = tagsViewModel.filteringTags.collectAsState().value
+  val tagsSuggestions = tagsViewModel.tagsSuggestions.collectAsState()
+  val stringQuery = tagsViewModel.query.collectAsState()
+
+  val expanded = remember { mutableStateOf(false) }
+
+  var showDialog by remember { mutableStateOf(false) }
+  var isModified by remember { mutableStateOf(false) }
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:70]: You sometimes set this var to true, but never actually use it. Is this intentional ?

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+  val defaultDate = calendar.time
+
+  val birthdate = remember { mutableStateOf(DateUtils.formatDate(defaultDate)) }
+  val selectedDate = remember { mutableStateOf(defaultDate) }
+  var isDateValid = remember { mutableStateOf(true) }
+
+  var selectedGender by remember { mutableStateOf<Gender?>(null) }
+
+  // Add form validation
+  fun isFormValid(): Boolean {
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:90]: As suggested by Botond, we might want to force the user to select at least x tags (you can create a variable MIN_REQUIRED_TAGS that we can modify then), and so you should check that the user has selected at least this minimum.

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:1]: Global nitpick: there are a lot of magic numbers on this file. Use fixed vars instead.

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+                      Profile(
+                          uid = user.uid,
+                          name = fullName.value,
+                          gender = selectedGender!!, // Add this line
+                          birthDate =
+                              Timestamp(
+                                  Calendar.getInstance()
+                                      .apply {
+                                        time = selectedDate.value!!
+                                        set(Calendar.HOUR_OF_DAY, 0)
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:274]: I think you have already set hour/minute/seconds/milliseconds on the pickedDate variable, so no need to set it again.

**Review by louiscrc — CHANGES_REQUESTED** [review]
Globally a very nice PR, essential in the user path.
I'll request changes just for the magic numbers, and so that you can check my remarks on certain portions of code.

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+                                        set(Calendar.MILLISECOND, 0)
+                                      }
+                                      .time),
+                          catchPhrase = catchphrase.value,
+                          description = description.value,
+                          tags = selectedTags,
+                          fcmToken = fcmToken)
+                  profilesViewModel.addNewProfile(newProfile)
+                  // Upload profile picture if selected
+                  profilesViewModel.validateAndUploadProfilePicture(context)
```
Comment (Kapestian) [inline app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:286]: Bug: to use the validateAndUploadProfilePicture you must set `selectedProfile` of `profilesViewModel` to the profile you just created, see why below:

image

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt (hunk 1 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:104]: Yes I'll change that!

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt (hunk 2 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:70]: No I will remove showDialog

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt (hunk 3 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:90]: As for now this isn't necessary and the team didn't agree on this dynamic

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:1]: Yes you are right I will fix that

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt (hunk 4 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/ProfileCreation.kt:274]: yep!

**Review by louiscrc — APPROVED** [review]
Nice PR ! LGTM

**Pull request merged**